osu.Desktop/LegacyIpc/LegacyTcpIpcProvider.cs

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Linq;
+using Newtonsoft.Json.Linq;
+using osu.Framework.Logging;
+using osu.Framework.Platform;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.Legacy;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Catch;
+using osu.Game.Rulesets.Mania;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Osu;
+using osu.Game.Rulesets.Taiko;
+
+namespace osu.Desktop.LegacyIpc
+{
+    /// <summary>
+    /// Provides IPC to legacy osu! clients.
+    /// </summary>
+    public class LegacyTcpIpcProvider : TcpIpcProvider
+    {
+        private static readonly Logger logger = Logger.GetLogger("legacy-ipc");
+
+        /// <summary>
+        /// Invoked when a message is received from a legacy client.
+        /// </summary>
+        public new Func<object, object> MessageReceived;
+
+        public LegacyTcpIpcProvider()
+            : base(45357)
+        {
+            base.MessageReceived += msg =>
+            {
+                try
+                {
+                    logger.Add($"Processing legacy IPC message...");
+                    logger.Add($"\t{msg.Value}", LogLevel.Debug);
+
+                    var legacyData = ((JObject)msg.Value).ToObject<LegacyIpcMessage.Data>();
```

Comment:
Any reason for using `JObject` everywhere rather than parsing the full object type? Some issue with matching types due to namespace differences?

Reply:
I need to look into this again. I had a response ready but I'm actually not sure how valid it is. The short explanation is that yes it's due to namespace differences and even with matching namespaces `Type.GetType()` wasn't able to find the type, but I need to go back and check if I was doing something wrong.

Reply:
Right, I remember now.

So the problem is that `IpcProvider` (the osu!framework base class) serialises types as `AssemblyQualifiedName` in order to make things work.
Since we're dealing with different assemblies, one would think we could just serialise the namespace + type name and do some sort of broad lookup, HOWEVER... `Type.GetType()` fails to find the type because it can only do this sort of broad lookup if the requested type is part of the currently-executing assembly (i.e. osu!framework, NOT the calling assembly).

> [`Type.GetType()`](https://docs.microsoft.com/en-us/dotnet/api/system.type.gettype?view=net-6.0#System_Type_GetType_System_String_)
>
> `typeName` String 
>
>The assembly-qualified name of the type to get. See AssemblyQualifiedName. If the type is in the currently executing assembly or in mscorlib.dll/System.Private.CoreLib.dll, it is sufficient to supply the type name qualified by its namespace.

So to get around this, `LegacyIpcMessage` exists which always serialises/deserialises the `Value` object as an `object` type which is part of the core library. Json.NET will deserialise this as a `JObject`, but we know that this object internally represents the `LegacyIpcMessage.Data` object.
Then, the `MessageType` contained in this object allows us to do the type matching of requests/responses ourselves.

I'll add some comments around to hopefully explain this better...

Reply:
I've added an explanation, hopefully it clears things up enough?

Reply:
Seems better, yep.